core/src/main/java/com/linecorp/armeria/common/RequestContextAwareMinimalStage.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+public class RequestContextAwareMinimalStage<T> implements CompletionStage<T> {
+
+    private final RequestContextAwareCompletableFuture<T> delegate;
+
+    RequestContextAwareMinimalStage(RequestContextAwareCompletableFuture<T> delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public <U> CompletionStage<U> thenApply(Function<? super T, ? extends U> fn) {
+        throw new UnsupportedOperationException();
```

Comment:
I think you have to implement all the methods that throw an `UnsupportedOperationException`:

```java
return new RequestContextAwareMinimalStage(
        (RequestContextAwareCompletableFuture<U>) delegate.thenApply(fn));
```